airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -155,29 +188,43 @@ protected void acceptTracked(final AirbyteMessage message) throws Exception {
            */
         }
       }
-      /*
-       * TODO: (ryankfu) after record has added and time has been met then to see if time has elapsed then
-       * flush the buffer
-       *
-       * The reason that this is where time component should exist here is primarily due to #acceptTracked
-       * is the method that processes each AirbyteMessage. Method to call is
-       * bufferingStrategy.flushWriter(stream, streamBuffer)
-       */
     } else if (message.getType() == Type.STATE) {
       stateManager.addState(message);
     } else {
       LOGGER.warn("Unexpected message: " + message.getType());
     }
+    periodicBufferFlush();
   }
 
   /**
-   * After marking states as committed, emit the state message to platform then clear state messages
-   * to avoid resending the same state message to the platform
+   * After marking states as committed, return the state message to platform then clear state messages
+   * to avoid resending the same state message to the platform. Also updates the next time a buffer
+   * flush should occur since it is deterministic that when this method is called all data has been
+   * successfully committed to destination
    */
   private void markStatesAsFlushedToDestination() {
     stateManager.markPendingAsCommitted();
     stateManager.listCommitted().forEach(outputRecordCollector);
     stateManager.clearCommitted();
+    nextFlushDeadline = Instant.now().plus(bufferFlushFrequency);
+  }
+
+  /**
+   * Periodically flushes buffered data to destination storage when exceeding flush deadline. Also
+   * resets the last time a flush occurred
+   */
+  private void periodicBufferFlush() {
+    // When the last time the buffered has been flushed exceed the frequency, flush the current
+    // buffer before receiving incoming AirbyteMessage
+    if (Instant.now().isAfter(nextFlushDeadline)) {
+      LOGGER.info("Periodic buffer flush started");
+      try {
+        bufferingStrategy.flushAll();
+        markStatesAsFlushedToDestination();
+      } catch (final Exception e) {
+        LOGGER.error("Periodic buffer flush failed", e);
```

Comment:
@ryankfu how do you feel about actually making this rethrow `e` instead of just logging it? a lot of these errors seem to be de facto nontransient, so we might as well just fail fast. e.g. https://github.com/airbytehq/oncall/issues/1751, where we run into a `state does not exist / not authorized` thing

or alternatively, maybe we can somehow hold on to the periodic flush exception + rethrow it if the sync fails? But that sounds kind of complicated, and potentially misleading in other ways

Reply:
I'm fine with re-throwing this Exception. I can spin up this as a quick PR or if you have one up feel free to tag me as a reviewer. Would probably go with the former solution since failing fast is more ideal because if we're not able to commit the buffered data it will result in data loss so retrying from the last saved STATE message will ensure data consistency